internal/devbox/providers/nixcache/nixcache.go

```diff
@@ -2,91 +2,201 @@ package nixcache
+	if err != nil {
+		return redact.Errorf("get path to current devbox executable: %s", err)
+	}
+	sudo, err := exec.LookPath("sudo")
+	if err != nil {
+		return redact.Errorf("get path to sudo executable: %s", err)
+	}
+	path, err := p.rootAWSConfigPath()
+	if err != nil {
+		return err
+	}
+
+	// Rename the .aws directory in case it already exists. We should
+	// improve this to be more careful with existing ~root/.aws/configs, but
+	// this seems rare enough that it should be okay for the initial
+	// implementation.
+	dir := filepath.Dir(path)
+	_ = os.Rename(dir, dir+".bak") // ignore errors for non-existent dir
+	_ = os.Mkdir(dir, 0o755)       // ignore errors for dir exists (don't os.MkdirAll the home directory)
+
+	config, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, fs.FileMode(0o644))
 	if err != nil {
-		return NixCacheConfig{}, err
+		return err
 	}
+	defer config.Close()
 
-	checkIfUserCanAddSubstituter(ctx)
+	// TODO(gcurtis): it would be nice to use a non-default profile if
+	// https://github.com/NixOS/nix/issues/5525 ever gets fixed.
+	_, err = fmt.Fprintf(config, `# This file was generated by Devbox.
+# Any overwritten configs can be found in the .aws.bak directory.
 
-	return NixCacheConfig{
-		URI: binCacheResponse.NixBinCacheUri,
-		Credentials: types.Credentials{
-			AccessKeyId:  aws.String(binCacheResponse.GetNixBinCacheCredentials().GetAccessKeyId()),
-			SecretKey:    aws.String(binCacheResponse.GetNixBinCacheCredentials().GetSecretKey()),
-			SessionToken: aws.String(binCacheResponse.GetNixBinCacheCredentials().GetSessionToken()),
-		},
-	}, nil
+[default]
+# sudo as the configured user so that their cached credential files have the
+# correct ownership.
+credential_process = %s -u %s -i %s cache credentials
+`, sudo, username, exe)
+	if err != nil {
+		return err
+	}
+	return config.Close()
+}
+
+func (p *Provider) sudoConfigureRoot(ctx context.Context, username string) error {
+	// TODO(gcurtis): save the user's response so that we don't pester them
+	// every time if it's a no.
+	prompt := &survey.Confirm{
+		Message: "Devbox requires root to configure the Nix daemon to use your organization's private cache. Allow sudo?",
+	}
+	ok := false
+	if err := survey.AskOne(prompt, &ok); err != nil {
+		return err
+	}
+	if !ok {
+		return nil
+	}
+
+	exe, err := os.Executable()
+	if err != nil {
+		return redact.Errorf("cannot determine path to current devbox executable: %s", err)
+	}
+
+	cmd := exec.CommandContext(ctx, "sudo", exe, "cache", "configure", "--user", username)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to relaunch with sudo: %w", err)
+	}
+	return nil
+}
+
+// Credentials fetches short-lived credentials that grant access to the user's
+// private cache.
+func (p *Provider) Credentials(ctx context.Context) (Credentials, error) {
+	cache := filecache.New[Credentials]("devbox/credentials")
```

Comment:
nix, should we call this cache `devbox/providers/nixcache` ?  Same for URI

Reply:
Also, after using this API, I realize it's weird for the templatizing to happen on `New`. Maybe we rename `New` to `ForType`

`filecache.ForType[Foo](name)`